query-algebrizer/src/cc.rs

```diff
@@ -0,0 +1,795 @@
+                        &PatternValuePlace::Variable(_) =>
+                            DatomsTable::AllDatoms,       // TODO: check types.
+                        &PatternValuePlace::Constant(NonIntegerConstant::Text(_)) =>
+                            DatomsTable::AllDatoms,
+                        _ =>
+                            DatomsTable::Datoms,
+                    }),
+            &PatternNonValuePlace::Entid(id) =>
+                schema.attribute_for_entid(id)
+                      .and_then(|attribute| self.table_for_attribute_and_value(attribute, value)),
+            &PatternNonValuePlace::Ident(ref kw) =>
+                self.attribute_for_ident(schema, kw)
+                    .and_then(|attribute| self.table_for_attribute_and_value(attribute, value)),
+        }
+    }
+
+    /// Produce a (table, alias) pair to handle the provided pattern.
+    /// This is a mutating method because it mutates the aliaser function!
+    /// Note that if this function decides that a pattern cannot match, it will flip
+    /// `is_known_empty`.
+    fn alias_table<'s, 'a>(&mut self, schema: &'s Schema, pattern: &'a Pattern) -> Option<SourceAlias> {
+        self.table_for_places(schema, &pattern.attribute, &pattern.value)
+            .map(|table| SourceAlias(table, (self.aliaser)(table)))
+    }
+
+    fn get_attribute<'s, 'a>(&self, schema: &'s Schema, pattern: &'a Pattern) -> Option<&'s Attribute> {
+        match pattern.attribute {
+            PatternNonValuePlace::Entid(id) =>
+                schema.attribute_for_entid(id),
+            PatternNonValuePlace::Ident(ref kw) =>
+                self.attribute_for_ident(schema, kw),
+            _ =>
+                None,
+        }
+    }
+
+    fn get_value_type<'s, 'a>(&self, schema: &'s Schema, pattern: &'a Pattern) -> Option<ValueType> {
+        self.get_attribute(schema, pattern).map(|x| x.value_type)
+    }
+}
+
+/// Expansions.
+impl ConjoiningClauses {
+
+    /// Take the contents of `bindings` and generate inter-constraints for the appropriate
+    /// columns into `wheres`.
+    /// For example, a bindings map like
+    ///
+    /// ```clojure
+    ///   {?foo [:datoms12.e :datoms13.v :datoms14.e]}
+    /// ```
+    ///
+    /// produces additional constraints:
+    ///
+    /// ```clojure
+    /// [[:= :datoms12.e :datoms13.v] [:= :datoms12.e :datoms14.e]]
+    /// ```
+    pub fn expand_bindings(&mut self) {
+        for cols in self.bindings.values() {
+            if cols.len() > 1 {
+                let ref primary = cols[0];
+                let secondaries = cols.iter().skip(1);
+                for secondary in secondaries {
+                    // TODO: if both primary and secondary are .v, should we make sure
+                    // the type tag columns also match?
+                    // We don't do so in the ClojureScript version.
+                    self.wheres.push(ColumnConstraint::EqualsColumn(primary.clone(), secondary.clone()));
+                }
```

Comment:
As discussed on Slack, I do think there's a separation of this algorithm into pipeline stages that might make some of these transformations clearer.